### DISCUSSION TITLE
How can I use Private module references in practice?

### DISCUSSION THREAD

**Discussion by hariscodes** [dagger/dagger#8433] [Help] [answered]
First off, I'm very grateful for the work done to get #7708 merged. Thanks to all who contributed to getting that out the door.

That being said, I'm confused on how I can actually use the feature since there's no documentation around it and I don't know how much of the discussion on that PR is part of the actual implementation vs forward-thinking conversations. 

Context: I have a private module hosted on GitHub that I'd like to run as part of a GitHub Actions pipeline. I'm currently doing this using a Git Submodule, which I'd prefer not to use if possible.

My questions are:
* What do I need to do on my GHA runner/in my workflow in terms of authentication setup? Can I use a PAT, or do I need to use SSH?
* How should I construct the module ref?

Thanks very much!

**Comment by grouville** [comment, marked as answer]
Hi @hariscodes, docs was just updated: https://docs.dagger.io/manuals/user/remote-repositories, module ref is detailed there 🙏

**Reply by grouville** [reply]
Currently only authentication via SSH refs is possible, PAT support is ongoing

**Reply by hariscodes** [reply]
Thank you!

**Comment by skycaptain** [comment]
Thank you for updating the documentation. While Dagger 0.14.0 allows me to access private remote modules via HTTPS on my local machine, I still cannot access the same module from a GitLab CI pipeline. Thus, it would be helpful if we could also add some examples for private modules on the “[https://docs.dagger.io/integrations/ci”](https://docs.dagger.io/integrations/ci%E2%80%9D) site.

For context: I have a private module on our self-hosted GitLab instance. While I can access the module from my local computer (using either SSH or HTTPS with a custom credential helper), I cannot access the same module from a GitLab CI pipeline using a `CI_JOB_TOKEN`. Interestingly, the same token works fine with a regular `git clone`, like this:

```
$ git clone https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.my.company/my/private/module.git
Cloning into '[redacted]'...
```

With a credential helper set up (I tried both `glab auth git-credential` and `store`), I can even git-clone without basic auth in the URL, like this:

```yaml
$ git clone https://gitlab.my.company/my/private/module.git
Cloning into '[redacted]'...
```

However, none of this works with Dagger:

```
# or: dagger call -m https://gitlab.my.company/my/private/module.git@main
# or: dagger call -m gitlab.my.company/my/private/module.git@main
$ dagger call -m https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.my.company/my/private/module.git@main
...
8   : loading module
9   :   finding module configuration
10  :     moduleSource(refString: "https://gitlab-ci-token:[MASKED]@gitlab.my.company/my/private/module.git@main"): ModuleSource!
11  :       cache request: git://gitlab.my.company/my/private/module.git#main
11  :       cache request: git://gitlab.my.company/my/private/module.git#main ERROR [0.1s]
11  :       ! failed to fetch remote https://xxxxx@gitlab.my.company/my/private/module.git: git error: exit status 128
stderr:
remote: HTTP Basic: Access denied. If a password was provided for Git authentication, the password was incorrect or you're required to use a token instead of a password. If a token was provided, it was either incorrect, expired, or improperly scoped. See https://gitlab.my.company/help/topics/git/troubleshooting_git.md#error-on-git-fetch-http-basic-access-denied
fatal: Authentication failed for 'https://gitlab.my.company/my/private/module.git/'
10  :     moduleSource ERROR [0.2s]
10  :     ! failed to resolve git src to commit: failed to load cache key: failed to fetch remote https://xxxxx@gitlab.my.company/my/private/module.git: git error: exit status 128
stderr:
remote: HTTP Basic: Access denied. If a password was provided for Git authentication, the password was incorrect or you're required to use a token instead of a password. If a token was provided, it was either incorrect, expired, or improperly scoped. See https://gitlab.my.company/help/topics/git/troubleshooting_git.md#error-on-git-fetch-http-basic-access-denied
fatal: Authentication failed for 'https://gitlab.my.company/my/private/module.git/'
...
```

**Reply by skycaptain** [reply]
After further debugging, I discovered that while a PAT works fine from a GitLab CI pipeline, a `CI_JOB_TOKEN` does not. This raises questions about the required permissions and recommended setup for GitLab CI, since `CI_JOB_TOKEN` has `read_repository` access and git-clone operations work perfectly with the same `CI_JOB_TOKEN` on the same repository from the same pipeline.

**Reply by skycaptain** [reply]
Duplicate of https://github.com/dagger/dagger/issues/8989
